### ISSUE TITLE
BitGravity hls live stream stops when currentTime equals duration

### ISSUE THREAD

**Issue by dtniland** [mangui/flashls#480]
@mangui 

We are running into an issue where a live stream always eventually stops in the flashls player.  It seems that sooner or later the currentTime will catch up with the duration and emit a PLAYBACK_COMPLETE event even though the stream has not really stopped.  The code that seems to be triggering the event comes from the HLSTimeTrait.  The _mediaTimeHandler kicks off the duration call which then causes the base TimeTrait to stop playback.

Here's the repro: http://www.flashls.org/latest/examples/osmf/GrindPlayer.html?src=http%3A%2F%2Fcnbclive.live-s.cdn.bitgravity.com%2Fcdn-live-dvr%2F_definst_%2Fcnbclive%2Fus-multiscreen%2Fus-mbr.smil%2Fplaylist.m3u8%253Fe%253D0%2526h%253D469b6053480444d28b791ac67b76889a%2526bgsecuredir%253D1%2526HLS

**Comment by mangui** [comment]
weird, I rechecked the whole logic, and flashls cannot fire PLAYBACK_COMPLETE on a live playlist
is it something specific to OSMF ?

**Comment by dtniland** [comment]
It seems to be the OSMF TimeTrait that fires the event.  It's blindly comparing the currentTime to the duration and quitting if they are equal.  Perhaps there needs to be code in the live stream handling to prevent the duration from equaling the currentTime, or perhaps the TimeTrait can be overridden to handle this?  

Here's the code that's being hit in the timeTrait, which comes from the _mediaTimeHandler in your HLSTimeTrait code.  It's hitting that signalComplete function.

protected final function setCurrentTime(value:Number):void
        {
            if (!isNaN(value))
            {
                // Don't ever let the currentTime exceed the duration.
                if (!isNaN(_duration))
                {
                    value = Math.min(value, _duration);
                }
                else
                {
                    value = 0;
                }
            }

```
        if (    _currentTime != value
            &&  !
                    (   isNaN(_currentTime)
                    &&  isNaN(value)
                    )
            )
        {
            currentTimeChangeStart(value);

            var oldCurrentTime:Number = _currentTime;
            _currentTime = value;

            currentTimeChangeEnd(oldCurrentTime);

            if (currentTime == duration && currentTime > 0)
            {
                signalComplete();
            } 
        }
    }
```

**Comment by mangui** [comment]
Ok then we just need to lower pos by 10ms   if pos === duration and stream type is live in HlsTimeTrait

If live
  pos = Math.min(pos,duration-0.01);

**Comment by dtniland** [comment]
yeah, it doesn't seem like 10 miliseconds would make much of a difference to anything else.  

We are also waiting on issue #473 so we can update flashls to the latest.

Thanks,
Daniel

**Comment by dtniland** [comment]
I'll make a fork and see if the fix works.

**Comment by dtniland** [comment]
I've put in the fix and it seems to work.  We'll run it through our QA next week to make sure there aren't any side effects then I'll make a pull request for you.

**Comment by dtniland** [comment]
Hmm, the fix didn't turn out to work very well.  There seems to be a timing issue where making the "gap" between the currentTime and the duration doesn't catch the issue all of the time.  In the timeTrait's setCurrentTime and setDuration calls, it compares the value not against the other value that is coming through in the current event, but against the value from the prior event, which will have already been set to the wrong value.

It seems like something unexpected is happening in the tick before the setDuration is hit.  I'll try to look at it again when I have time -- but it might not be for several days.

**Comment by dtniland** [comment]
All right I figured it out.  The stream was occasionally getting into a state where the currentTime and the duration were the same, and the first fix was good for that.  But often, on the very next tick of the mediaTime, the duration and time would be equal to each other again, but both would be a smaller number than before.  I assume that there is a bit of variation as the stream "slides" through its window.

When the duration of the current tick was less than the currentTime of the previous tick, the end condition would fire in the TimeTrait.  I fixed this by constraining the duration to never be smaller than its previous value.

**Comment by mangui** [comment]
understood. wouldn't the following be even more accurate ?

```
        /** Update playback position/duration **/
        private function _mediaTimeHandler(event : HLSEvent) : void {
            var newDuration : Number = event.mediatime.duration;
            var newPosition : Number = Math.max(0, event.mediatime.position);
            // ensure that new duration is at least 10ms greater than current position
            setDuration(Math.max(newDuration,position+0.01));
            // ensure that new position is at least 10ms smaller than current duration
            setCurrentTime(Math.min(newPosition,duration-0.01));
        };
```

**Comment by dtniland** [comment]
It would be dangerous to put it there, since we don't know whether the stream is live or not.  If we put this logic on a normal stream, it might never end.  Instead I put it in the StreamBuffer._checkBuffer function

`
                var lastDur:Number = isNaN(_lastDuration) ? 0 : _lastDuration;
                duration = _hls.type == HLSTypes.LIVE ? Math.max(pos + .01, _playlistDuration, lastDur) : _playlistDuration;
`

The only gotcha here is that the live stream will not now shrink its duration in the mediaTime update even if the actual media has shrunk a little.  This seems to be fine on the bitGravity live stream as the duration seems to hover around a single value +-2 seconds or so, and seeking past the duration seems to start the stream at the beginning of the window.  If there are live media that shrink the duration more aggressively, then this could become a noticeable problem.  This could be fixed by keeping a third duration value that would allow it to get back to its actual duration on the next tick, but that gets pretty complicated so I haven't put it in.

**Comment by mangui** [comment]
Hi @dtniland 
as it is an OSMF specific issue i would prefer that the fix remains in the OSMF glue.
we can eventually add if (hls.type === HLSTypes.LIVE) around 

```
        /** Update playback position/duration **/
        private function _mediaTimeHandler(event : HLSEvent) : void {
            var newDuration : Number = event.mediatime.duration;
            var newPosition : Number = Math.max(0, event.mediatime.position);
            if (hls.type === HLSTypes.LIVE) {
              // ensure that new duration is at least 10ms greater than current position
               setDuration(Math.max(newDuration,position+0.01));
              // ensure that new position is at least 10ms smaller than current duration
              setCurrentTime(Math.min(newPosition,duration-0.01));
            } else {
            // business as usual
             setDuration(newDuration);
             setCurrentTime(newPosition);
           }
        };
```

**Comment by dtniland** [comment]
I'm afraid that we've discovered that the issue seems deeper than we'd thought.  Even with the code that prevents the stream from ending via the TimeTrait, we're seeing that the stream starts to behave erratically soon after it starts playing.  The live stream can get into a state where it bounces between stream indexes and every time it changes the index, it points to a different location in the stream, making the live media appear to bounce back and forth.  Often the video freezes while the audio jumps around until eventually everything stops working; both video and audio are frozen.  When it freezes, fragments are still loading into the player even though nothing is playing.

When we play the same stream in the safari video tag, it also gets a little choppy, but the playback never stops. 

The trace log doesn't show anything strange other than the currentIndex getting set between different streams.

**Comment by mangui** [comment]
Having choppyness in Safari is a bad sign stream-wise... could you share debug logs ?

**Comment by dtniland** [comment]
All right, I dumped the debug2 logs
[bitgravitytrace1.txt](https://github.com/mangui/flashls/files/120032/bitgravitytrace1.txt)
[bitgravitytrace2.txt](https://github.com/mangui/flashls/files/120031/bitgravitytrace2.txt)
[bitgravitytraceStartIssue.txt](https://github.com/mangui/flashls/files/120030/bitgravitytraceStartIssue.txt)

traceStartIssue.txt is the trace from near the very beginning of the point where the video freezes.  I hope I got the point where it actually stops playing, but I can't tell for sure.

trace1.txt is the trace I get when the video has stopped but the audio is still working.

trace2.txt is after everything is dead.

All of this trace is just a small sampling of a huge amount of trace logs that are zooming past.

**Comment by dtniland** [comment]
Also, we have reports that bitgravity is looking at the issue too from their end.  They may have been tweaking something when we saw the stuttering video because we haven't seen it before or since.  We have seen the video consistently freeze, however, once we got it to stop ending prematurely.

**Comment by jlacivita** [comment]
Hi @mangui.  Myself and @dtniland have developed what seems like a more reliable fix for the flashls OSMF HLSTimeTrait.  We're testing it now, but it's been running for 18 hours (and still going), as opposed to an average of 3 hours w/out the fix.

I'll send a pull request in a bit:
https://github.com/jlacivita/flashls/blob/dev/src/org/mangui/osmf/plugins/traits/HLSTimeTrait.as

**Comment by mangui** [comment]
clever

**Comment by jlacivita** [comment]
okay, our stream has been running for almost 24 hours.  i'm going to send a pull request now.

**Comment by jlacivita** [comment]
We're going to go ahead and use our fork for now, but here's the pull request when you get time:
https://github.com/mangui/flashls/pull/487

**Comment by jlacivita** [comment]
This is fixed here:
https://github.com/mangui/flashls/blob/dev/src/org/mangui/osmf/plugins/traits/HLSTimeTrait.as#L41

I think we can close this issue.
